cmd/gpu_levelzero/zes.c

```diff
@@ -0,0 +1,441 @@
+        return ZE_RESULT_ERROR_NOT_AVAILABLE;
+    }
+
+    if (count > 1) {
+        fprintf(stderr, "more than one zes driver detected, using first one\n");
+    }
+
+    count = 1;
+
+    zes_driver_handle_t handle;
+    res = zesDriverGet(&count, &handle);
+    if (res != ZE_RESULT_SUCCESS) {
+        return res;
+    }
+
+    count = 0;
+    res = zesDeviceGet(handle, &count, NULL);
+    if (res != ZE_RESULT_SUCCESS) {
+        return res;
+    }
+
+    if (count == 0) {
+        return ZE_RESULT_ERROR_NOT_AVAILABLE;
+    }
+
+    zes_handles = calloc(count, sizeof(zes_device_handle_t));
+
+    res = zesDeviceGet(handle, &count, zes_handles);
+    if (res != ZE_RESULT_SUCCESS) {
+        free(zes_handles);
+
+        return res;
+    }
+
+    zes_handles_count = count;
+
+    bdf_addresses = (struct device_info*) calloc(count,sizeof(struct device_info));
+    if (bdf_addresses == NULL) {
+        free(zes_handles);
+
+        return ZE_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    // Iterate over the devices and store their info into the cache array
+    for (uint32_t i = 0; i < count; ++i) {
+        zes_device_handle_t dev_h = zes_handles[i];
+
+        zes_pci_properties_t pci_props;
+        if (zesDevicePciGetProperties(dev_h, &pci_props) != ZE_RESULT_SUCCESS) {
+            continue;
+        }
+
+        zes_pci_address_t* addr = &pci_props.address;
+
+        snprintf(bdf_addresses[i].bdf, sizeof(bdf_addresses[i].bdf),
+            "%04x:%02x:%02x.%x",
+            addr->domain, addr->bus, addr->device, addr->function
+        );
+    }
+
+    device_enumerated = true;
+
+    return res;
+}
+
+zes_device_handle_t retrieve_handle_for_bdf(char* bdf_address)
+{
+    zes_device_handle_t handle = 0;
+
+    for (uint32_t i = 0; i < zes_handles_count; ++i) {
+        struct device_info* di = &bdf_addresses[i];
+
+        if (strncmp(bdf_address, di->bdf, sizeof(di->bdf)) == 0) {
+            handle = zes_handles[i];
+            break;
+        }
+    }
+
+    return handle;
+}
+
+bool is_integrated(zes_device_handle_t handle)
+{
+    ze_result_t res = ZE_RESULT_SUCCESS;
+
+    zes_device_ext_properties_t ext = {
+        .stype = ZES_STRUCTURE_TYPE_DEVICE_EXT_PROPERTIES,
+    };
+    zes_device_properties_t props = {
+        .stype = ZES_STRUCTURE_TYPE_DEVICE_PROPERTIES,
+        .pNext = &ext,
+    };
+
+    if (res = zesDeviceGetProperties(handle, &props), res == ZE_RESULT_SUCCESS) {
+        if (ext.flags & ZES_DEVICE_PROPERTY_FLAG_INTEGRATED) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+/// @brief Retrieves memory amount for a specific device with bdf address
+/// @param bdf_address
+/// @return memory amount for the device
+uint64_t zes_device_memory_amount(char* bdf_address, uint32_t* error)
+{
+    if (getenv("UNITTEST") != NULL) {
+        return 0;
+    }
+
+    if (verbosity_level >= verbosity_debug) {
+        fprintf(stderr, "Retrieve memory size for %s\n", bdf_address);
+    }
```

Comment:
Uh, this check repeated so many times that IMHO there should be helper(s) for them.  For example:
```
static void log(log_level_t level, const char *msg) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s\n", msg);
    }
}

static void log_number(log_level_t level, const char *msg, const char *value) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s: %s\n", msg, value);
    }
}

static void log_string(log_level_t level, const char *msg, const char *value) {
    if (verbosity_level >= level) {
        fprintf(stderr, "%s: %s\n", msg, value);
    }
}
```
```suggestion
    log_string(LOG_DEBUG, "Retrieve memory size for", bdf_address);
```

Or if you'd rather have single log function, switch to using `va_args`.